lib/macroni/Translation/Kernel/KernelVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+    return std::nullopt;
+  }
+
+  auto loc = m_view.location(rcu_assign_pointer);
+  auto rty = m_bld.void_type();
+  auto params = m_rcu_assign_pointer_params.at(rcu_assign_pointer);
+  auto p = params.p;
+  auto v = params.v;
+  auto visitor = vast::cg::default_stmt_visitor(m_bld, m_view, scope);
+  auto p_op = m_view.visit(p, scope);
+  auto v_op = m_view.visit(v, scope);
+  auto p_result = p_op->getOpResult(0);
+  auto v_result = v_op->getOpResult(0);
+
+  return m_bld.create<RCUAssignPointer>(loc, rty, p_result, v_result);
+}
+
+std::optional<vast::operation>
+kernel_visitor::visit_rcu_access_pointer(const vast::cg::clang_stmt *stmt,
+                                         vast::cg::scope_context &scope) {
+  auto rcu_access_pointer = clang::dyn_cast<clang::StmtExpr>(stmt);
+  if (!rcu_access_pointer) {
+    return std::nullopt;
+  }
+
+  if (!m_rcu_access_pointer_to_p.contains(rcu_access_pointer)) {
+    return std::nullopt;
+  }
+
+  auto loc = m_view.location(rcu_access_pointer);
+  auto p = m_rcu_access_pointer_to_p.at(rcu_access_pointer);
+  vast::cg::default_stmt_visitor visitor(m_bld, m_view, scope);
+  auto rty = m_view.visit(p->getType(), scope);
+  auto p_op = m_view.visit(p, scope);
+
+  return m_bld.create<RCUAccessPointer>(loc, rty, p_op->getOpResult(0));
+}
+
+std::optional<vast::operation>
+kernel_visitor::visit_rcu_replace_pointer(const vast::cg::clang_stmt *stmt,
+                                          vast::cg::scope_context &scope) {
+  auto rcu_replace_pointer = clang::dyn_cast<clang::StmtExpr>(stmt);
+  if (!rcu_replace_pointer) {
+    return std::nullopt;
+  }
+
+  if (!m_rcu_replace_pointer_to_params.contains(rcu_replace_pointer)) {
+    return std::nullopt;
+  }
+
+  auto loc = m_view.location(rcu_replace_pointer);
+  auto params = m_rcu_replace_pointer_to_params.at(rcu_replace_pointer);
+  vast::cg::default_stmt_visitor visitor(m_bld, m_view, scope);
+  auto rty = m_view.visit(params.rcu_ptr->getType(), scope);
+  auto rcu_ptr_op = m_view.visit(params.rcu_ptr, scope);
+  auto ptr_op = m_view.visit(params.ptr, scope);
+  auto c_op = m_view.visit(params.c, scope);
+
+  return m_bld.create<RCUReplacePointer>(loc, rty, rcu_ptr_op->getOpResult(0),
+                                         ptr_op->getOpResult(0),
+                                         c_op->getOpResult(0));
+}
+
+template <typename AttrT>
+void annotate_op_with_attrs_in_text(vast::operation op, std::string_view text,
+                                    AttrT attr) {
+  if (text.contains(AttrT::getMnemonic())) {
+    op->setAttr(AttrT::getMnemonic(), attr);
+  }
+}
+
+template <typename AttrT, typename... Rest>
+void annotate_op_with_attrs_in_text(vast::operation op, std::string_view text,
+                                    AttrT attr, Rest... rest) {
+  if (text.contains(AttrT::getMnemonic())) {
+    op->setAttr(AttrT::getMnemonic(), attr);
+  }
+  annotate_op_with_attrs_in_text(op, text, rest...);
```

Comment:
Hm, I prefer it
```
template< H, ... Tail > ...

if constexpr ( sizeof ... (Tail) != 0)
  recurse;
```
to avoid that code duplication.

Reply:
Cool